RealmSwift/Tests/ThreadSafeReferenceTests.swift

```diff
@@ -330,3 +330,480 @@ class ThreadSafeReferenceTests: TestCase {
+            try realm.write {
+                realm.create(SwiftStringObject.self, value: ["A"])
+                realm.create(SwiftStringObject.self, value: ["B"])
+                realm.create(SwiftStringObject.self, value: ["C"])
+                realm.create(SwiftStringObject.self, value: ["D"])
+            }
+        }
+
+        let realm = try! Realm()
+        @ThreadSafe var results = realm.objects(SwiftStringObject.self)
+        dispatchSyncNewThread {
+            guard let results = results else {
+                return XCTFail("no results")
+            }
+            results.indices.forEach { idx in
+                XCTAssertEqual(results[idx].stringCol, values[idx])
+            }
+        }
+        @ThreadSafe var swiftStringObject = results!.first
+        dispatchSyncNewThread {
+            guard let swiftStringObject = swiftStringObject else {
+                return XCTFail("no results")
+            }
+
+            XCTAssertEqual(swiftStringObject.stringCol, "A")
+        }
+    }
+
+    func mutateStringCol(@ThreadSafe stringObj: SwiftStringObject?) {
+        guard let stringObj = stringObj else {
+            XCTFail("no object found")
+            return
+        }
+        try! stringObj.realm!.write {
+            stringObj.stringCol = "after"
+        }
+    }
+
+    func testThreadSafeFunctionArgument() {
+        let realm = try! Realm()
+        @ThreadSafe var stringObj = try! realm.write {
+            realm.create(SwiftStringObject.self, value: ["stringCol": "before"])
+        }
+        dispatchSyncNewThread {
+            XCTAssertEqual(stringObj!.stringCol, "before")
+            self.mutateStringCol(stringObj: stringObj)
+        }
+        realm.refresh()
+        XCTAssertEqual(stringObj!.stringCol, "after")
+    }
+
+    func testThreadSafeUnmanagedArgument() {
+        let stringObj = SwiftStringObject(value: ["stringCol": "before"])
+
+        dispatchSyncNewThread {
+            self.assertThrows(self.mutateStringCol(stringObj: stringObj), reason: "Only managed objects may be wrapped as thread safe.")
+        }
+    }
+
+    func testThreadSafeMultipleDispatch() {
+        let realm = try! Realm()
+        @ThreadSafe var obj = try! realm.write {
+            realm.create(SwiftStringObject.self, value: ["stringCol": "before"])
+        }
+        let backgroundQueue = DispatchQueue(label: "background", attributes: .concurrent)
+        let mainEx = expectation(description: "executes first block")
+        let backgroundEx = expectation(description: "executes second block")
+
+        backgroundQueue.async {
+            XCTAssertEqual(obj?.stringCol, "before")
+            try! obj?.realm?.write {
+                obj?.stringCol = "middle"
+            }
+            mainEx.fulfill()
+        }
+        backgroundQueue.async {
+            try! obj?.realm?.write {
+                obj?.stringCol = "after"
+            }
+            backgroundEx.fulfill()
+        }
```

Comment:
`concurrentPerform(iterations:execute:)` with something like 100 iterations would exercise this a bit better.